sources/web/datalab/jupyter.ts

```diff
@@ -64,14 +66,19 @@ var templates: common.Map<string> = {
  */
 var appSettings: common.Settings;
 
-function pipeOutput(stream: NodeJS.ReadableStream, port: number, error: boolean) {
+function pipeOutput(server: JupyterServer, stream: NodeJS.ReadableStream, port: number, error: boolean) {
   stream.setEncoding('utf8');
+  var listeningMessage = `The Jupyter Notebook is running at: http://127.0.0.1:${port}`;
+
   stream.on('data', (data: string) => {
     // Jupyter generates a polling kernel message once every 3 seconds
     // per kernel! This adds too much noise into the log, so avoid
     // logging it.
 
     if (data.indexOf('Polling kernel') < 0) {
+      if (data.indexOf(listeningMessage) >= 0) {
```

Comment:
Yes. Another reason I don't like it is that the logic is not in the same function. If we do wait_for_port_free --> start_jyputer_process -> wait_for_port_used (which may not be the safest either, I agree), we could put all logic inside the same function, and is easier to understand.